### PR TITLE
Add Button Click tracking plugin

### PR DESCRIPTION
**Pull request by greg-el**
This PR adds the Button Click tracking plugin that now ships by default with the v4 JS tracker.

A dropdown has been added in when the user is configuring a tag with `enableButtonClickTracking` tag type selected, allowing the user to pick between an allowlist or denylist. I've left out the [function filter](https://docs.snowplow.io/docs/collecting-data/collecting-from-own-applications/javascript-trackers/web-tracker/tracking-events/button-click/#filter-function) since I couldn't really think of a way of doing it that wasn't at least slightly unpleasant.

Also bumped the name/desc to v4.

### CODE REVIEW

File: template.tpl [hunk 1]
```diff
@@ -87,6 +87,10 @@ ___TEMPLATE_PARAMETERS___
       {
         "displayValue": "[Custom Command]",
         "value": "customCommand"
+      },
+      {
+        "value": "enableButtonClickTracking",
+        "displayValue": "Button Click Tracking"
```
Comment (jethron) [inline template.tpl]: Can we keep this list alphabetized (except for `[Custom Command]` as before?

File: template.tpl [hunk 2]
```diff
@@ -1071,6 +1080,48 @@ ___TEMPLATE_PARAMETERS___
...
+    "subParams": [
+      {
+        "type": "SELECT",
+        "name": "buttonClickTrackingFilterChoice",
+        "displayName": "Filter Type",
+        "macrosInSelect": false,
+        "selectItems": [
+          {
+            "value": "allowlist",
+            "displayValue": "Allowlist"
```
Comment (jethron) [inline template.tpl]: The display values could probably just be "Allow" and "Deny", but I don't think it's a big deal.

File: template.tpl [hunk 3]
```diff
@@ -2023,6 +2085,23 @@ switch (data.eventType) {
...
+    
+  case 'enableButtonClickTracking':
+    commandName = data.eventType;
+    
+    // `data.buttonClickTrackingFilterChoice` will always be set,
+    // being "allowlist" by default
+    const choice = data.buttonClickTrackingFilterChoice;
+    const filters = data.buttonClickTrackingFilters
+      .split(",")
+      .map(f => f.trim());
```
Comment (jethron) [inline template.tpl:2105]: Doesn't this mean we will always use a `{filter: {allowlist: [""]}}` default and never track anything unless an actual comprehensive list of classes is provided? I don't think [`classes.some((className) => filter.allowlist.includes(className))`](https://github.com/snowplow/snowplow-javascript-tracker/blob/e835323adfe8ec7b8ea084a33fe3a7f34efa1df8/plugins/browser-plugin-button-click-tracking/src/util.ts#L21) will ever succeed with that input?

Should we do this bit later and guard it in a check on `data.buttonClickTrackingFilters.length` or something?

File: template.tpl [hunk 4]
```diff
@@ -1071,6 +1080,48 @@ ___TEMPLATE_PARAMETERS___
...
+        ],
+        "simpleValueType": true,
+        "subParams": [
+          {
+            "type": "TEXT",
+            "name": "buttonClickTrackingFilters",
+            "displayName": "Classes",
+            "simpleValueType": true,
+            "help": "A comma-seperated list of HTML class names for filtering.",
+            "valueHint": "htmlClass, secondHtmlClass"
```
Comment (jethron) [inline template.tpl]: Does this need the "Allow empty strings"/`canBeEmptyString` option to not be `undefined` in the code later and cause a JS error?

File: template.tpl [hunk 5]
```diff
@@ -2023,6 +2085,23 @@ switch (data.eventType) {
...
     }
     break;
+    
+  case 'enableButtonClickTracking':
+    commandName = data.eventType;
+    
+    // `data.buttonClickTrackingFilterChoice` will always be set,
+    // being "allowlist" by default
+    const choice = data.buttonClickTrackingFilterChoice;
+    const filters = data.buttonClickTrackingFilters
```
Comment (jethron) [inline template.tpl]: Since this can be configured as a variable we might want to check if it's an array already unless we want to mandate strings only.

**Review by jethron — COMMENTED** [review]
I think a filter function would probably get sandboxed and not work properly anyway (it would get no actual DOM element which limits its utility). So I think that's fine (it's also consistent with the Link Tracking config which also doesn't accept a function).

File: template.tpl (hunk 2 above)
Comment (greg-el) [reply template.tpl]: Allow/Deny sounds a bit better to me too, updated.

File: template.tpl (hunk 4 above)
Comment (greg-el) [reply template.tpl]: I've handled `undefined`/general falsey in the JS, `undefined` is treated as no filter options being set for the button tracking now.

File: template.tpl (hunk 3 above)
Comment (greg-el) [reply template.tpl:2105]: Yeah, I've fiddled with this a bit now. The command/parameters are only set if the value for filter is `undefined` (no filter set), or we can parse a sensible non-zero-length array of strings out of the value provided.

**Review by jethron — APPROVED** [review]
LGTM

**Pull request merged**